GPII-1526: Port native addons to nan

javihernandez opened the pull request:
Hey @amb26, 

I've ported our native modules to nan, so this is ready for you to start reviewing it.
The only missing piece is the packageKit addon, which is being ported by @klown and will end up in this branch for you to review too.

javihernandez committed:
GPII-1526: Add nan as a devDependency

javihernandez committed:
GPII-1526: Poting gsettings addon to Nan

javihernandez committed:
GPII-1526: Removed useless tests.js file under gsettingsBridge folder

javihernandez committed:
GPII-1526: Ported ALSA addon to nan

javihernandez committed:
GPII-1526: Ported XRandR addon to nan

amb26 commented on gpii/node_modules/gsettingsBridge/nodegsettings/nodegsettings.cc:
```diff
@@ -15,119 +16,119 @@
-  args[1]->ToString()->WriteAscii(key);
-  if (args[2]->IsBoolean()) {
-    status = g_settings_set_boolean(settings,key,args[2]->BooleanValue());
+  settings = g_settings_new(*String::Utf8Value(schema));
```
This would be a good opportunity to fix our long-standing bug due to never deallocating this object. By my reading - we should call g_object_unref, but I may misunderstand - http://stackoverflow.com/questions/2848273/should-a-g-object-new-have-a-matching-g-object-unref
What do you think?

klown replied:
Yes.
Certainly I did in the Gnome Control Panel for the Magnifier, e.g.:
https://git.gnome.org/browse/gnome-control-center/tree/panels/universal-access/zoom-options.c#n438

klown replied:
Actually g_clear_object (&settings) would be better:
https://developer.gnome.org/gobject/unstable/gobject-The-Base-Object-Type.html#g-clear-object

javihernandez replied:
@amb26, @klown,

**ACTUALLY YES**, so I've also added some "deallocating" code in both gsettings and packageKit modules.

klown replied:
@javihernandez 
I wrote:

> Actually g_clear_object (&settings) would be better.

I've updated the packagekit add-on in that regard (nodepackagekit.cc), and pushed it this morning.  If you want to keep it in your GPII-1526 branch, you'll have to pull my latest in.

javihernandez replied:
@klown, 

I've just merged in your changes

klown committed:
GPII-1526: Ported PackageKit addon to nan.

javihernandez committed:
GPII-1526: Ported PackageKit addon to nan

* klown/GPII-1526:
  GPII-1526: Ported PackageKit addon to nan.

javihernandez committed:
GPII-1526: Freeing some g_objects

klown committed:
GPII-1526: Ported PackageKit addon to nan.

Used g_clear_object() instead of the conditional g_object_unref().
Cleaned up white space.

klown committed:
Merge branch 'javi-GPII-1526' into GPII-1526

Conflicts:
	gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc

javihernandez committed:
GPII-1526: Use g_object_clear as a replacement of g_object_unref

* klown/GPII-1526:
  GPII-1526: Ported PackageKit addon to nan.

Conflicts:
	gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc

javihernandez committed:
GPII-1526: Updated reference to universal in AcceptanceTests.js

klown committed:
Merge branch 'javi-GPII-1526' into GPII-1526
Attempt #2.
Conflicts:
	gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc

javihernandez committed:
GPII-1526: Replace fluid's makeChangeApplier with makeHolderChangeApplier

klown committed:
Merge branch 'javi-GPII-1526' into GPII-1526

amb26 commented on gpii/node_modules/gsettingsBridge/nodegsettings/nodegsettings.cc:
```diff
@@ -15,119 +16,131 @@
+  Handle<Array> togo = Array::New(isolate, size);
   for (i = 0; keys[i]; i++) {
-    togo->Set(i,String::New(keys[i]));
+    togo->Set(i,String::NewFromUtf8(isolate, keys[i]));
```
I don't see this strategy for constructing strings advertised in the public NAN docs - where did you find the recommendation?

What I see in the docs is this: https://github.com/nodejs/nan/blob/master/doc/string_bytes.md#api_nan_encode which seems to avoid the need for the bizarre "Isolate" reference...

javihernandez replied:
Right,

My fault, I saw the weird "Isolate" thing in a few examples from nodejs official addon examples, which wasn't using Nan. In fact, now I'm realizing that would be better to use `Nan::New(keys[i]).ToLocalChecked()` instead, since internally we're making use of a `MaybeLocal` type. Is that correct?

klown replied:
@javihernandez 
I used Nan::Encode() to construct strings without the need for Isolate, as @amb26 suggested.
Look inside makePac() of nodepackagekit.cc
https://github.com/klown/linux/blob/GPII-1526/gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc#L85

I haven't addressed other uses of Isolate in that code yet, though.

klown replied:
I wrote:

> I haven't addressed other uses of Isolate in that code yet, though.

I have removed the use of Isolate in nodepackagekit.cc when allocating v8 Arrays and a Boolean.  There is still a use in one case with a ThrowError.  I'll address that on Mon.

klown replied:
I wrote:

> >    I haven't addressed other uses of Isolate in that code yet, though.
> 
> I have removed the use of Isolate in nodepackagekit.cc when allocating v8 Arrays and a
> Boolean. There is still a use in one case with a ThrowError. I'll address that on Mon.

Done.  There are no more instances of Isolate in nodepackagekit.cc.

amb26 commented on gpii/node_modules/gsettingsBridge/nodegsettings/nodegsettings.cc:
```diff
@@ -15,119 +16,131 @@
   }
   else if (g_variant_type_equal(type,G_VARIANT_TYPE_INT32)) {
-    return scope.Close(Int32::New(g_variant_get_int32(variant)));
+    info.GetReturnValue().Set(Int32::New(isolate, g_variant_get_int32(variant)));
```
Docs suggest the use of Nan::New here - https://github.com/nodejs/nan/blob/master/doc/new.md#api_nan_new - where did you find this approach?

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -84,7 +84,7 @@
         var currentSettingsInProfile = userSettings.profiles[userSettings.general.activeProfile[PROFILE_ID]];
 
-        var applier = fluid.makeChangeApplier(currentSettingsInProfile, {
+        var applier = fluid.makeHolderChangeApplier(currentSettingsInProfile, {
```
I am quite sure this will fail. If it passes, this implies a serious fault in the test cases which should be fixed - I imagine that the fault is that you don't attempt to test any removal of settings - although I can't imagine that the tests which try to verify that the user's original settings have been restored will pass - did you try them?

https://github.com/GPII/universal/blob/master/tests/platform/linux/linux-orca-testSpec.js#L25

Either way we have a serious problem that our test cases can't be testing what we hope they are testing

javihernandez replied:
Good question,

ACTUALLY the tests are passing, so let me check what's going on here.

amb26 commented on gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc:
```diff
@@ -30,12 +38,10 @@ static gboolean DEBUG = FALSE;
-    string->WriteUtf8(buffer);
-    return buffer;
+    Local<String> string = value.As<String>();
+    return (*String::Utf8Value(string));
```
Hard to tell whether we have lost a memory leak or gained one here :)
Perhaps someone can walk me through the expectation here - we believe that the Utf8Value object will be destructed on the stack, the "string" will be destructed on the stack shortly afterwards, deallocating its buffer, which has been copied into the returned string in the meantime?
http://stackoverflow.com/a/1032671/5041298 reminds us that every construction of a std::string will do a heap allocation as it copies the buffer. This suggests that we once had a memory leak - can we be clear that we don't have one any more?
It looks like there are some Nan utilities we should use to make this clearer and more portable rather than relying on V8 primitives, even if we succeed in convincing ourselves that this code is valid.

klown replied:
@amb26, I think you've walked yourself thought it pretty much :)

You're right. Using the nan utilities would make this clearer.  I'll fix it.

klown committed:
GPII-1526: Port of PackageKit addon to nan.

Modified ut8StringFromValue() to use Nan string functions.

Addresses amb26's comment:
https://github.com/amb26/linux/pull/2/#discussion_r47282147

klown committed:
GPII-1526: PackageKit nan port:  strings.

Modified makePac() to eliminate using Isolate object for string
construction.

Addresses amb26's comment:
https://github.com/amb26/linux/pull/2/#discussion_r47279041

klown committed:
GPII-1526: PackageKit nan port:  Isolate

Reomved usage of Isolate with respect to allocating a v8 Array
and Boolean.

klown committed:
GPII-1526: PackageKit nan port.

Modified makePac() to use a loop to get various properties of an
individual package.

klown committed:
GPII-1526: PackageKit nan port:  Remove Isolate.

Removed use of Isolate by the ThrowError() within performAction().
Added unit test of this ThrowError() by calling performAction() such
that it fails.

Addresses amb26's comment:
https://github.com/amb26/linux/pull/2/#discussion_r47279041

javihernandez committed:
GPII-1526: Removed Isolate

Also, updated the code to use Local handles when possible.

javihernandez committed:
GPII-1526: Use console.log instead of the deprecated util.puts

javihernandez committed:
GPII-1526: Remove Isolate from PackageKit addon

* klown/GPII-1526:
  GPII-1526: PackageKit nan port:  Remove Isolate.
  GPII-1526: PackageKit nan port.
  GPII-1526: PackageKit nan port:  Isolate
  GPII-1526: PackageKit nan port:  strings.
  GPII-1526: Port of PackageKit addon to nan.

javihernandez commented:
Updated, and merged in latest changes from @klown.
Now looking at the issue with orca.

javihernandez committed:
GPII-1526: Reworked orca's settingsHandler

Updating tests to ensure that we're leaving the settings file as it was.
Removed the auto-creation of the Orca's config file.

amb26 commented on gpii/node_modules/gsettingsBridge/nodegsettings/nodegsettings.cc:
```diff
@@ -15,119 +16,132 @@
-    ThrowException(Exception::Error(String::New("Need to implement reading that value type")));
-    return scope.Close(Undefined());
+    ThrowError(New("Need to implement reading that value type").ToLocalChecked());
+    info.GetReturnValue().Set(Nan::False());
```
This line will be unreachable. If our cleanup hasn't been properly set up by now (i.e. via a destructor) it will be too late.

javihernandez replied:
That's true, I guess that we're ok just by throwing the error, isn't it?

amb26 commented on gpii/node_modules/gsettingsBridge/nodegsettings/nodegsettings_tests.js:
```diff
@@ -65,26 +64,26 @@ var testIntegerValues = function () {
 var testReadingDoubleValues = function () {
     var ret = gsettings.get_gsetting(keyScreenmag, "mag-factor");
-    util.puts("The mag-factor is: " + ret);
+    console.log("The mag-factor is: " + ret);
```
All these "tests" which have been changed don't actually make any assertions

javihernandez replied:
Yes,

should I just add some assertions or do you want me to rewrite all these tests to use jqUnit and so on?

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -53,129 +41,98 @@
+            } else if (value === undefined) {
+                type = "DELETE";
+            } else {
+                type = null;
```
This is still an "ADD"

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -53,129 +41,98 @@
-            };
-            change.type = settingVal === undefined ? "DELETE" : "ADD";
-            applier.fireChangeRequest(change);
+    gpii.orca.settingsHandler.onCreate = function (that) {
```
Interesting approach, but we should leave a note that this does not definitely resolve the problem. I don't see why the settings file should have been constructed onCreate of this component when we weren't sure it was constructed much latter when the settings handler was actually used. We should restore the comment that this needs to be converted to an asynchronous settings handler.

javihernandez replied:
Ok,

I've reworked this to be a promise-based initialization of the component. I added the same Orca's settings file initialization process we had before, so now we're trying to create the settings file at component creation time. Let me know what do you think about it.

javihernandez replied:
@amb26, my only question is,

how about the get and set methods? Will they catch the rejected promise when they are trying to resolve model.settings and throw the corresponding error, or do I need to add something else?

amb26 replied:
No, this design is still not correct. The promises must be returned from the get/set methods - see our sample at https://github.com/amb26/universal/blob/GPII-1318/gpii/node_modules/testing/src/Integration.js#L235

amb26 commented on gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc:
```diff
@@ -13,14 +13,22 @@ You may obtain a copy of the License at
 #include <string>
 
 using namespace v8;
+using v8::FunctionTemplate;
```
I'd prefer to do without these using directives so we can be crystal clear when we are using a v8 primitive and when a Nan one

klown replied:
Okay.

amb26 commented on gpii/node_modules/xrandr/nodexrandr/nodexrandr.cc:
```diff
@@ -39,6 +39,13 @@
 #endif
 
 using namespace v8;
+using v8::FunctionTemplate;
```
See above

amb26 commented on gpii/node_modules/xrandr/nodexrandr/nodexrandr.cc:
```diff
@@ -39,6 +39,13 @@
+using Nan::New;
+using Nan::Set;
+
 using namespace std;
```
See http://stackoverflow.com/questions/1452721/why-is-using-namespace-std-in-c-considered-bad-practice - I think we should axe all "using"s

klown replied:
Agreed.

amb26 commented on gpii/node_modules/alsa/nodealsa/nodealsa.cc:
```diff
@@ -14,18 +15,22 @@
 #include <math.h>
 #include <alsa/asoundlib.h>
 
 using namespace v8;
```
Global comment - please remove all "using" directives

amb26 commented on gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc:
```diff
@@ -29,13 +37,20 @@ static gboolean DEBUG = FALSE;
-    return buffer;
+    std::basic_string<gchar> result;
+    Nan::Utf8String nanUtf8 (value);
+    if (nanUtf8.length() != 0) {
```
What happens if this check is removed?

klown replied:
According to the [Nan documentation](https://github.com/nodejs/nan/blob/master/doc/v8_misc.md#nanutf8string), if the Nan::Utf8String conversion fails, then length() returns zero, but using the \* operator (used in the next line) gives a NULL result and not an empty string.  Thus, if length() is zero, there's nothing to append, and don't bother trying.

Unless I'm reading the Nan docs wrong.

amb26 replied:
Well, this is a subtlety well worthy of being documented in the code - please could you add a comment which refers to those docs together with this explanation - cheers

klown replied:
Done.

amb26 commented on gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc:
```diff
@@ -149,45 +179,46 @@ Handle<Value> searchPackage(const Arguments& args) {
-  pk_package_sack_sort(sack, PK_PACKAGE_SACK_SORT_TYPE_NAME);
-  array = pk_package_sack_get_array(sack);
+  sack = pk_results_get_package_sack(pkResults);
+  pk_package_sack_sort (sack, PK_PACKAGE_SACK_SORT_TYPE_NAME);
```
Please remove all spaces which occur after the name of a function in a function call. It would be great if we could locate some kind of linting task we can apply to these files, but in the meantime we will have to do it by eye.

klown replied:
A holdover from my GNOME programming days.  The [GNOME style guide](https://developer.gnome.org/programming-guidelines/stable/c-coding-style.html.en) requires spaces after function names.  Is there a gpii c/c++ style guide? ;-)

I'll remove the spaces.

amb26 replied:
Our commit standards for JS only defer to Crockford - we should follow whatever of these conventions are valid in C/C++ - http://javascript.crockford.com/code.html - plus whatever can be applied from our JSHint file - https://github.com/fluid-project/infusion/blob/master/.jshintrc - e.g. 4 spaces for indents, etc. We should start to assemble coding conventions for C/C++ since we are writing more and more of it (e.g. avoidance of "using", etc.) and add them to our page: https://wiki.fluidproject.org/display/fluid/Coding+and+Commit+Standards

amb26 commented on gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc:
```diff
@@ -29,13 +37,20 @@ static gboolean DEBUG = FALSE;
+    if (nanUtf8.length() != 0) {
+      result.append (*nanUtf8);
+    }
+    return (result);
```
Unnecessary parentheses - or if they are necessary for some truly arcane reason, please comment why :)

klown replied:
Oooo, a challenge.  Here's an [arcane reason](https://stackoverflow.com/questions/4762662/are-parentheses-around-the-result-significant-in-a-return-statement#answer-25615981).  But, it's not applicable here, so I'll remove the parentheses.

amb26 replied:
The discovery that there _might_ be an arcane reason adds even further impetus to remove them :)

klown committed:
GPII-1526: PackageKit nan port: response to Antranig's review.

- Removed all 'using' statements.
- Removed all spaces after function names.
- Removed parentheses from 'return' statements.

See:
https://github.com/amb26/linux/pull/2/#discussion_r47279041

klown committed:
GPII-1526: XRandr nan port: addressing Antranig's review.

- Removed all 'using' statements.
- Removed all spaces after function names.

See:
https://github.com/amb26/linux/pull/2/#discussion_r47279041

klown commented:
@javihernandez While modifying the nodexrandr.cc code to address @amb26's comments I noticed two calls to `exit()` in the [`getDisplays()`](https://github.com/javihernandez/linux/blob/GPII-1526/gpii/node_modules/xrandr/nodexrandr/nodexrandr.cc#L161) function.  Is that good idea, since it would cause the node process itself to stop?  Wouldn't it be better to return an empty array in this case?  The situation is where XOpenDisplay() fails, and the add-on can't then get the properties of the display(s).

javihernandez committed:
GPII-1526: Addressing Antranig's review

* klown/GPII-1526:
  GPII-1526: XRandr nan port: addressing Antranig's review.
  GPII-1526: PackageKit nan port: response to Antranig's review.

javihernandez committed:
GPII-1526: Removed "using" statements

javihernandez committed:
GPII-1526: Deprecated g_settings_list_keys

Using g_settings_schema_list_keys as suggested in the documentation.
See:
 * https://bugzilla.gnome.org/show_bug.cgi?id=740308
 * https://developer.gnome.org/gio/2.46/GSettings.html#g-settings-list-keys

javihernandez commented:
> @javihernandez While modifying the nodexrandr.cc code to address @amb26's comments I noticed two calls to exit() in the getDisplays() function. Is that good idea, since it would cause the node process itself to stop? Wouldn't it be better to return an empty array in this case? The situation is where XOpenDisplay() fails, and the add-on can't then get the properties of the display(s).

Right @klown,

what about throwing an error?

klown commented:
@javihernandez 

> what about throwing an error?

I have no opinion.  What is the advantage?  What happens when the xrandr add-on is called during a full run of the gpii and the error is thrown?

Also, how do we test this?

javihernandez committed:
GPII-1526: Added some asserts to nodegsettings_tests.js

javihernandez committed:
GPII-1526: Addressing some more comments from Antranig

amb26 commented:
Clearest way to signal failure is to move over to the promise-style API for settings handlers and return a rejection. We should do this for orca too. Currently our only sample of these is in test cases: https://github.com/amb26/universal/blob/GPII-1318/gpii/node_modules/testing/src/Integration.js#L236

javihernandez committed:
GPII-1526: Added a promise-based approach to make Orca configurable

By configurable we understand that the Orca settings file exists and that we
have read/write access to it.

javihernandez committed:
GPII-1526: Removing "using" statements from ALSA's add-on

klown committed:
GPII-1526: PackageKit nan port: added comment re: NanUtf8 strings.

Adding comment about avoiding NULL.
https://github.com/amb26/linux/pull/2#discussion_r60114798

klown committed:
GPII-1526: XRandr nan port: throws errors instead of exit().

Modified getDisplays() to throw an error if the code fails
to get the display or its screen.  Also, minor code corrections.
https://github.com/amb26/linux/pull/2#issuecomment-210550789

klown commented on gpii/node_modules/packagekit/nodepackagekit/nodepackagekit.cc:
```diff
@@ -273,25 +294,31 @@ Handle<Value> performAction(const Arguments& args) {
+      Nan::ThrowError(
+          "You have to provide the action to be performed, either 'install', " 
+          "'update' or 'remove'");
   }
```
@javihernandez There is a potential memory leak here.  The `task` and `package_ids` were allocated above, and are properly freed just below this throw block.  However, if the throw is executed, they are not freed.

Patch coming up.

klown committed:
GPII-1526: PackageKit nan port: freed data structures.

Modified performAction():  For unsupported actions, free and
NULL-ify allocated data structures prior to throwing an error.
https://github.com/amb26/linux/pull/2#discussion_r60930273

klown committed:
GPII-1526: Xrandr add-on nan port: improved unit test.

The test of setting the screen resolution to impossible dimensions
was erroneously using the physical size of the display as the
impossible resolution.  Changed to simply use a resolution of
(-1 ,-1).

https://github.com/amb26/linux/pull/2

klown committed:
GPII-1526: Xrandr add-on nan port: fixed memory leak.

javihernandez committed:
GPII-1526: Some more memory leak fixes from Joseph

* klown/GPII-1526:
  GPII-1526: Xrandr add-on nan port: fixed memory leak.
  GPII-1526: Xrandr add-on nan port: improved unit test.
  GPII-1526: PackageKit nan port: freed data structures.
  GPII-1526: XRandr nan port: throws errors instead of exit().
  GPII-1526: PackageKit nan port: added comment re: NanUtf8 strings.

klown commented:
@javihernandez @amb26 A heads up:  GPII needs to support Fedora-23/glib-2.46.
I merged all of @javihernandez's changes with mine and re-built everything.  `nodegsettings.cc` does not compile because it depends on glib-2.46 (for `g_settings_schema_list_keys()`) and Fedora-22 only goes as high as glib-2.44.  I can upgrade to Fedora-23 to solve this, no problem; but then so does GPII as a whole (or some other distro that includes glib-2.46).

javihernandez commented:
> @javihernandez @amb26 A heads up: GPII needs to support Fedora-23/glib-2.46.
> I merged all of @javihernandez's changes with mine and re-built everything. nodegsettings.cc does not compile because it depends on glib-2.46 (for g_settings_schema_list_keys()) and Fedora-22 only goes as high as glib-2.44. I can upgrade to Fedora-23 to solve this, no problem; but then so does GPII as a whole (or some other distro that includes glib-2.46).

oh, I see, and sorry for that. Do you want me to add backwards compatibility or just do a "heads up" communication to the rest of the team? (mostly thinking about our provisioning system)

amb26 commented:
I think a headsup is fine. I don't see a need for us to depend on a version of Fedora that will receive EOL before we produce the APCP release.

amb26 commented on gpii/node_modules/packagekit/nodepackagekit/nodepackagekit_test.js:
```diff
@@ -155,6 +155,17 @@ var fluid = require("universal"),
         return matchPkg;
     };
 
+    pkTests.testPerformActionFailure = function() {
```
Two linting failures in this function - please ensure to run "grunt lint" before issuing pull - see @the-t-in-rtf work on git commit hooks

klown replied:
Sorry, will fix these as well as the one below.

the-t-in-rtf replied:
Just in case, here's the archive of the note:

http://lists.idrc.ocad.ca/pipermail/architecture/2016-April/004051.html

It typically prevents me from committing badly formatted code a few times a week, totally worth it.

klown replied:
Thanks for the link, @the-t-in-rtf !
Aside:  as a test, I added the hook to my copy of universal, added a lint violation to gpii.js, and tried to commit it.  Nothing happened.  It turns out that universal's Gruntfile.js checks only the files under the "`gpii`", "`tests`", and "`examples`" directories.  Adding "`./*.js`" to the list fixed that and reported two errors in `gpii.js`, besides my cooked-up one.  Is there a reason for not linting the files in the top level directory?

the-t-in-rtf replied:
Not sure if that was directed at me.  My concern there would be cleanly avoid linting node_modules and other content that isn't part of the project.  It seems like your approach handles that.

klown replied:
@the-t-in-rtf That was not directed at you, specifically.  And, you are correct that we don't want to lint all the third-party material.  But, `gpii.js`, `index.js`, and possibly `package.json` are GPII specific, and they aren't being linted (and `gpii.js` shows lint errors).
However, this is a tangent to _this_ pull request.  I'll shut up now, and raise an appropriate JIRA for it.

amb26 commented on gpii/node_modules/packagekit/nodepackagekit/nodepackagekit_test.js:
```diff
@@ -228,6 +239,16 @@ var fluid = require("universal"),
         }
     });
 
+    jqUnit.test("Test performAction() with invalid argments", function() {
```
Lint failure here

Commit:
GPII-1526:  Fixed lint errors in nodepackagekit unit tests.

klown committed:
GPII-1774:  Display progress for PackageKit node add-on.

Set DEBUG flag to true so that progress is printed always.  Also,
tweaked progressCallback() to tidy up its output, and to flush
stdout to ensure that the percentages are printed immediately.

javihernandez committed:
GPII-1526: Orca's get/set methods now return promises

javihernandez committed:
GPII-1526: Joseph's fix for GPII-1774 and some minor linting

* klown/GPII-1526:
  GPII-1774:  Display progress for PackageKit node add-on.
  GPII-1526:  Fixed lint errors in nodepackagekit unit tests.

javihernandez commented:
@amb26 

I've included the latest changes from Joseph and fixed the implementation of Orca's get/set methods (they return promises only if the settings file doesn't exist, and tries to create it). One question, do the operations that require read/write access to files must be enclosed inside a try/catch block, and rejecting the promise if anything goes wrong?

amb26 commented:
Yes, they must!

javihernandez committed:
GPII-1526: Added try/catch blocks in case orca settings handler fails

javihernandez commented:
> Yes, they must!

Ok, done

javihernandez commented:
@amb26,

is there anything else do you want to see happening before accepting this?

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -78,104 +53,165 @@
+        };
+
+        if (exists) {
+            try {
```
Very repetitive. This logic should be captured in the getSettings method itself.
Also, the standard form for a rejection payload is
{
    isError: true,
    message: "Couldn't get settings: " + err
}

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -78,104 +53,165 @@
 
-        fs.writeFileSync(orcaSettingsFile,
-                         JSON.stringify(userSettings, null, 4));
+        var getSettings = function () {
```
This should be a public namespaced function as per our best practices - also it should incorporate the repeated logic for producing rejections

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -78,104 +53,165 @@
-
-            orcaSpawn.kill("SIGKILL");
+        } else {
+            var err = gpii.orca.settingsHandler.makeConfigurable();
```
Refactor makeConfigurable to return a promise and remove this repetitive logic

javihernandez committed:
GPII-1526: WIP promise-based orca settings handler

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -78,104 +53,206 @@
-                return;
-            }
-        }
+        var getSettings = that.getSettings(payload);
```
This can all just be 
    return that.getSettings(payload)

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -78,104 +53,206 @@
 
-        var returnValue = applySettings(profile);
-        returnObj = returnValue;
+        var setSettings = that.setSettings(payload);
```
This can all just be 
    return that.setSettings(payload)

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -78,104 +53,206 @@
+    gpii.orca.settingsHandler.getSettings = function (that, payload) {
+        var togo = fluid.promise();
+
+        var loadSettings = that.loadSettings();
```
I don't think it helps much to make the settingsHandler a modelComponent.
Instead, rewrite loadSettings so that it yields the actual settings in its resolve(), then this block can just be

```
    return fluid.promise.map(loadSettings, function (settings) {
        var profile = settings.general.activeProfile[PROFILE_ID];
```

etc.

amb26 commented on gpii/node_modules/orca/orcaSettingsHandler.js:
```diff
@@ -78,104 +53,206 @@
+        var checkIfFileGetsCreated = function () {
+            pass++;
+            if (pass === maxPass) {
+               orcaSpawn.kill("SIGKILL");
```
linting fails for indentation of this block.

javihernandez committed:
GPII-1526: Addressed last comments from Antranig wrt Orca

javihernandez committed:
GPII-1526: Use promise.map instead of .then when calling loadSettings
